File: backend/app/contract.py
"""Canonical per-turn JSON contract (verbatim from the ``json-contract`` skill).

This is the single source of truth for the shape every ``/chat`` turn returns. It is
later set as the PydanticAI orchestrator ``output_type``; for now the ``/chat`` stub
emits a type-valid placeholder. The nine fields, names, and constraints are
non-negotiable — do not add, rename, or drop fields.

Requirement: platform-scaffold-011 (define the canonical ``TurnOutput`` and
``GuardrailReport`` Pydantic models exactly per the constitution's per-turn JSON
contract).
"""

from pydantic import BaseModel, Field
from pydantic_extra_types.country import CountryAlpha2  # ISO 3166-1 alpha-2


class GuardrailReport(BaseModel):
    """Names of guardrails that triggered this turn (empty when clean)."""

    input: list[str] = Field(default_factory=list)
    output: list[str] = Field(default_factory=list)


class TurnOutput(BaseModel):
    reply: str = Field(..., description="User-facing answer, in active_lang")
    detected_lang: str = Field(
        ..., min_length=2, max_length=2, description="ISO 639-1 the user wrote in"
    )
    active_lang: str = Field(
        ...,
        min_length=2,
        max_length=2,
        description="Language the session is locked to (es|en|pt or fallback)",
    )
    lang_confidence: float = Field(
        ..., ge=0.0, le=1.0, description="Agreement score: LLM detected_lang vs lingua detector"
    )
    final_normalized_text: str = Field(
        ..., description="LLM-cleaned user text fused with resolved locale"
    )
    detected_country: CountryAlpha2 | None = Field(
        default=None, description="Fused geo-IP signal (ISO 3166-1 alpha-2)"
    )
    confidence_score: float = Field(
        ..., ge=0.0, le=1.0, description="Combined reconciliation confidence"
    )
    needs_review: bool = Field(
        default=False, description="True on low confidence / divergence / caught errors"
    )
    guardrails: GuardrailReport = Field(default_factory=GuardrailReport)

File: frontend/lib/contract.ts
/**
 * lib/contract.ts
 *
 * TypeScript mirror of the canonical per-turn JSON contract (9 fields, verbatim
 * names). The frontend is a read-only consumer — it never computes contract fields.
 *
 * Traces: frontend-shell-002 (session_id shape), frontend-shell-003 (POST body),
 *         frontend-shell-004 (render reply), frontend-shell-006 (non-contract error)
 */

// ---------------------------------------------------------------------------
// TurnOutput — the canonical per-turn contract
// All 9 field names are verbatim from the root CLAUDE.md specification.
// ---------------------------------------------------------------------------

export interface TurnOutput {
  /** User-facing answer from the orchestrator. */
  reply: string;
  /** ISO 639-1 code the user wrote in. */
  detected_lang: string;
  /** ISO 639-1 code the session is locked to (backend guarantees es|en|pt). */
  active_lang: string;
  /** Agreement score between LLM-detected language and lingua detector (0–1). */
  lang_confidence: number;
  /** LLM + API fused, locale-normalised input text. */
  final_normalized_text: string;
  /** Fused geo signal — ISO 3166-1 alpha-2 country code. */
  detected_country: string;
  /** Combined logic confidence score (0–1). */
  confidence_score: number;
  /** True when confidence is low, languages diverge, or errors occurred. */
  needs_review: boolean;
  /** Triggered guardrail names — never expose raw internals to students. */
  guardrails: { input: string[]; output: string[] };
}

// ---------------------------------------------------------------------------
// isTurnOutput — runtime type guard
//
// Validates every field's presence and type so a non-contract body from the
// server is detectable and can be converted to ApiError (req frontend-shell-006).
// ---------------------------------------------------------------------------

export function isTurnOutput(x: unknown): x is TurnOutput {
  if (typeof x !== "object" || x === null) return false;
  const o = x as Record<string, unknown>;

  if (typeof o.reply !== "string") return false;
  if (typeof o.detected_lang !== "string") return false;
  if (typeof o.active_lang !== "string") return false;
  if (typeof o.lang_confidence !== "number") return false;
  if (typeof o.final_normalized_text !== "string") return false;
  if (typeof o.detected_country !== "string") return false;
  if (typeof o.confidence_score !== "number") return false;
  if (typeof o.needs_review !== "boolean") return false;

  // Validate guardrails shape
  if (typeof o.guardrails !== "object" || o.guardrails === null) return false;
  const gr = o.guardrails as Record<string, unknown>;
  if (!Array.isArray(gr.input)) return false;
  if (!gr.input.every((v: unknown) => typeof v === "string")) return false;
  if (!Array.isArray(gr.output)) return false;
  if (!gr.output.every((v: unknown) => typeof v === "string")) return false;

  return true;
}

// ---------------------------------------------------------------------------
// ChatTurn — view-model union for the transcript
//
// One discriminated variant per actor: student message, assistant reply
// (carrying the full contract for ContractMeta rendering), or error turn.
// ---------------------------------------------------------------------------

export type ChatTurn =
  | { role: "student"; text: string }
  | { role: "assistant"; text: string; contract: TurnOutput }
  | { role: "error"; text: string };
